### PR TITLE
Simple predictions

### LINKED ISSUE #101: Predictions based on staypoints and regularly visited POIs
As a final feature for the first version, the app is supposed to predict when the user is about to leave for a regularly visited POI (see https://github.com/sitcomlab/simport-learning-app/issues/100). One way of doing this in a simple way could be setting up something like a schedule / timetable for the whole week that contains (for every hour or half our?) the most visited location (or an approximation if it) or POI. This lookup table is updated constantly and from time to time it is checked for regularities or patterns (e.g. daily or weekly cycles). If a pattern is detected, we can use it for a simple prediction that is then presented to the user.

This concept is still very rough around the edges and before implemented something like this it would make sense to invest a bit of time in background research.

### PR DESCRIPTION
**Pull request by felixerdy**
closes #101 

Features:
- Calculates the visit count for all POI inferences for each hour within a week
- Triggers notification if user is likely to visit a POI in the next hour
- Timetable chart in explore page

During review, please pay attention to the background fetch implementation. Maybe there is a more elegant way

### CODE REVIEW

File: src/app/shared-services/db/migrations.ts [hunk 1]
```diff
@@ -116,6 +116,35 @@ export const MIGRATIONS = [
   // add field 'start' to table points
   `ALTER TABLE points ADD COLUMN state TEXT`,
 
+  // add primary key 'id' to table inferendes
```
Comment (schrooom) [inline src/app/shared-services/db/migrations.ts:126]: These new migration-entries should append to the end of the list of migrations after the last current statement (`CREATE TABLE IF NOT EXISTS reverseGeocoding`). On devices with the latest state of the app, this way not all of the new migrations would execute.

File: src/app/shared-services/inferences/inference.service.ts [hunk 2]
```diff
@@ -142,16 +142,12 @@ export class InferenceService implements OnDestroy {
     }
   }
 
-  async generateUserInferencesWithDialog(callback: () => Promise<void>) {
+  protected async backgroundFuction(): Promise<void> {
```
Comment (schrooom) [inline src/app/shared-services/inferences/inference.service.ts:145]: Typo "fuction"

File: src/app/shared-services/background/AbstractBackgroundService.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+          })
+        }
+        this.backgroundService.backgroundState = BackgroundState.foreground
+
+        await this.internalBackgroundFuction(callback)
+      })
+    }
+  }
+
+  private async internalBackgroundFuction(
```
Comment (schrooom) [inline src/app/shared-services/background/AbstractBackgroundService.ts]: Typo "fuction"

File: src/app/shared-services/background/AbstractBackgroundService.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      this.backgroundService.backgroundState = BackgroundState.idle
+      if (callback !== undefined) await callback()
+    }
+  }
+
+  /**
+   * The background function to run.
+   * @param callback callback function
+   */
+  protected abstract backgroundFuction(): Promise<void>
```
Comment (schrooom) [inline src/app/shared-services/background/AbstractBackgroundService.ts]: Typo "fuction"

File: src/app/shared-services/timetable/timetable.service.spec.fixtures.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
+import { Inference } from 'src/app/model/inference'
+import { InferenceType } from '../inferences/engine/types'
+
+export const oneMondayFiveToNine = new Inference(
+  'inferenceId',
```
Comment (schrooom) [inline src/app/shared-services/timetable/timetable.service.spec.fixtures.ts:5]: The IDs `inferenceId` and `inferenceId2` are used (at least) twice each - is this intentional? (if it makes any difference at all)

File: src/app/shared-services/timetable/timetable.service.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import { NotificationType } from '../notification/types'
+import { TrajectoryService } from '../trajectory/trajectory.service'
+import { Timetable } from './timetable'
+
+@Injectable({
+  providedIn: 'root',
+})
+export class TimetableService extends AbstractBackgroundService {
+  protected foregroundInterval = 720
+  protected backgroundInterval = 15
```
Comment (schrooom) [inline src/app/shared-services/timetable/timetable.service.ts:17]: There are varying intervals in our background-services (`InferenceService` vs `TimetableService`) – is this a good idea, given that we only have one instance of background-fetch? (state here https://github.com/sitcomlab/simport-learning-app/issues/101#issuecomment-986581065)

**Comment by schrooom** [comment]
Looks very good so far - great idea to abstract the background-service! Some minor comments on the code as you see above.

Furthermore two points regarding the feature:
- Occasionally the "generating inferences"-dialog pops up and does not disappear by itself. I have to close and reopen the app to achieve that (just going back to home is sufficient, no need to fully shutdown the app) – we had this kind of behaviour before, I could take a look if you like.
- When using "Predict next visit", nothing is happening for me. I think some feedback to the user would be helpful. Where would potential results of this pop up either way?

File: src/app/shared-services/background/AbstractBackgroundService.ts (hunk 3 above)
Comment (felixerdy) [reply src/app/shared-services/background/AbstractBackgroundService.ts]: resolved in 06ee40f

File: src/app/shared-services/background/AbstractBackgroundService.ts (hunk 4 above)
Comment (felixerdy) [reply src/app/shared-services/background/AbstractBackgroundService.ts]: resolved in 06ee40f

File: src/app/shared-services/db/migrations.ts (hunk 1 above)
Comment (felixerdy) [reply src/app/shared-services/db/migrations.ts:126]: resolved in 18cf7d0

**Comment by felixerdy** [comment]
> * When using "Predict next visit", nothing is happening for me. I think some feedback to the user would be helpful. Where would potential results of this pop up either way?

resolved in 84f3f1b

Predictions will be highlighted on the map. When there is no prediction a toast will pop up

> * Occasionally the "generating inferences"-dialog pops up and does not disappear by itself. I have to close and reopen the app to achieve that (just going back to home is sufficient, no need to fully shutdown the app) – we had this kind of behaviour before, I could take a look if you like.

resolved in 5b8676c

File: src/app/shared-services/timetable/timetable.service.spec.fixtures.ts (hunk 5 above)
Comment (jbraese) [reply src/app/shared-services/timetable/timetable.service.spec.fixtures.ts:5]: The way the tests are set up right now only inferences with different IDs are passed to the timetable together, so it shouldn't make a difference.

File: src/app/shared-services/timetable/timetable.service.ts (hunk 6 above)
Comment (felixerdy) [reply src/app/shared-services/timetable/timetable.service.ts:17]: I'm not that sure about that as both intervals are still used in the corresponding `triggerBackgroundFunctionAsTask` and `triggerBackgroundFunctionAsFetch` functions to check whether the execution is in the correct schedule.

**Pull request merged**